// wood-mysql/src/mysql.js
// mysql操作方法类
// by YuRonghui 2018-7-9
//const Query = require('wood-querysql');
const Sequelize = require('sequelize');
//const { Util } = require('wood-util')();
//const { error, catchErr } = Util;
let dbs = {};

class Mysql {
  constructor(tbName, dbName = 'master') {
    this.tableName = tbName;
    this.dbName = dbName;
    if (dbs[this.dbName]) {
      this.db = dbs[this.dbName];
    } else {
      throw error('mysql failed: db=null');
    }
  }
  // 创建数据库连接池
  static async connect(opts, name = 'master') {
    if (dbs[name]) {
      console.error(`Mysql connect ${name} already exsit!`);
      return;
    }
    let { database, user, password, ...options } = opts;
    options.dialect = 'mysql';
    try {
      dbs[name] = new Sequelize(database, user, password, options);
    } catch (error) {
      console.log(error);
    }
    console.log('Mysql connected Successfull');
  }

  define(tableName, opts) {
    return this.getConn().define(tableName, opts);
  }

  query(sql, options) {
    return this.getConn().query(sql, options);
  }
  // 获取连接
  getConn(database) {
    return dbs[database || this.dbName];
  }
}

module.exports = Mysql;
